SpatialGDK/Source/Private/Interop/SnapshotManager.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SnapshotManager.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/GlobalStateManager.h"
+#include "Interop/SpatialReceiver.h"
+#include "SpatialConstants.h"
+#include "Utils/SchemaUtils.h"
+
+DEFINE_LOG_CATEGORY(LogSnapshotManager);
+
+using namespace improbable;
+
+void USnapshotManager::Init(USpatialNetDriver* InNetDriver, UGlobalStateManager* InGlobalStateManager)
+{
+	NetDriver = InNetDriver;
+	Receiver = InNetDriver->Receiver;
+	GlobalStateManager = InGlobalStateManager;
+}
+
+// WorldWipe will send out an expensive entity query for every entity in the deployment.
+// It does this by having an entity query for all entities that are not the GSM (workaround for not having the ability to make a query for all entities).
+// Once it has the response to this query, it will send deletion requests for all found entities and then one for the GSM itself.
+// Should only be triggered by the worker which is authoritative over the GSM.
+void USnapshotManager::WorldWipe(const USpatialNetDriver::ServerTravelDelegate& Delegate)
+{
+	UE_LOG(LogSnapshotManager, Warning, TEXT("World wipe for deployment has been triggered. All entities will be deleted."));
+
+	Worker_Constraint GSMConstraint;
+	GSMConstraint.constraint_type = WORKER_CONSTRAINT_TYPE_ENTITY_ID;
+	GSMConstraint.entity_id_constraint.entity_id = GlobalStateManager->GlobalStateManagerEntityId;
+
+	Worker_NotConstraint NotGSMConstraint;
+	NotGSMConstraint.constraint = &GSMConstraint;
+
+	Worker_Constraint WorldConstraint;
+	WorldConstraint.constraint_type = WORKER_CONSTRAINT_TYPE_NOT;
+	WorldConstraint.not_constraint = NotGSMConstraint;
+
+	Worker_EntityQuery WorldQuery{};
+	WorldQuery.constraint = WorldConstraint;
+	WorldQuery.result_type = WORKER_RESULT_TYPE_SNAPSHOT;
+
+	Worker_RequestId RequestID;
+	RequestID = NetDriver->Connection->SendEntityQueryRequest(&WorldQuery);
+
+	EntityQueryDelegate WorldQueryDelegate;
+	WorldQueryDelegate.BindLambda([this, Delegate](Worker_EntityQueryResponseOp& Op)
+	{
+		if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+		{
+			UE_LOG(LogSnapshotManager, Error, TEXT("SnapshotManager WorldWipe - World entity query failed: %s"), UTF8_TO_TCHAR(Op.message));
+		}
+
+		if (Op.result_count == 0)
+		{
+			UE_LOG(LogSnapshotManager, Error, TEXT("SnapshotManager WorldWipe - No entities found in world entity query"));
+		}
+
+		if (Op.result_count >= 1)
+		{
+			// Send deletion requests for all entities found in the world entity query.
+			DeleteEntities(Op);
+
+			// Also make sure that we kill the GSM.
+			NetDriver->Connection->SendDeleteEntityRequest(GlobalStateManager->GlobalStateManagerEntityId);
+		}
+
+		Delegate.ExecuteIfBound();
```

Comment:
Perhaps add a comment about how we've now cleared the world, and we're ready for the server to travel to the next/new level.